### ISSUE TITLE
[apps/speed.c OpenSSL 3.0.2] RSA verification error with OpenSSL speed test w/ async_jobs and pka engine that supports async_jobs

### ISSUE THREAD

**Issue by shihyic** [openssl/openssl#20589]

**Comment by mattcaswell** [comment]
You don't seem to have a provided a description for this issue. Please can you describe the problem that you are encountering.

**Comment by shihyic** [comment]
Ah, sorry. My description seems to be lost somewhere. :)

I am testing a PKA engine that supports ASYNC_jobs. With RSA tests, it involves in blinding which is to resolve a security risk. Currently every loopargs[] shares single EVP_PKEY rsa info. With this two different loopargs[] testing would update the blinding "keys" like rsa->A and rsa->Ai. It would interfere with each other. RSA signature tests will succeed but RSA verification would fail with errors:

RSA verify setup failure.  No RSA verify will be done.
20B0A5A4FFFF0000:error:0200008A:rsa routines:RSA_padding_check_PKCS1_type_1:invalid padding:crypto/rsa/rsa_pk1.c:75:
20B0A5A4FFFF0000:error:02000072:rsa routines:rsa_ossl_public_decrypt:padding check failed:crypto/rsa/rsa_ossl.c:680:

I solved this by creating a EVP_PKEY rsa for each loopsargs and it solve the problem.

Below is the git diff based on openssl-3.0.2

```
diff --git a/apps/speed.c b/apps/speed.c
index d252b41ca4..b90faf9103 100644
--- a/apps/speed.c
+++ b/apps/speed.c
@@ -2331,7 +2331,7 @@ int speed_main(int argc, char **argv)
             goto end;

     for (testnum = 0; testnum < RSA_NUM; testnum++) {
-        EVP_PKEY *rsa_key = NULL;
+        EVP_PKEY **rsa_pkeys = malloc(loopargs_len * sizeof(intptr_t));
         int st = 0;

         if (!rsa_doit[testnum])
@@ -2339,27 +2339,33 @@ int speed_main(int argc, char **argv)

         if (primes > RSA_DEFAULT_PRIME_NUM) {
             /* we haven't set keys yet,  generate multi-prime RSA keys */
-            bn = BN_new();
-            st = bn != NULL
-                && BN_set_word(bn, RSA_F4)
-                && init_gen_str(&genctx, "RSA", NULL, 0, NULL, NULL)
-                && EVP_PKEY_CTX_set_rsa_keygen_bits(genctx, rsa_keys[testnum].bits) > 0
-                && EVP_PKEY_CTX_set1_rsa_keygen_pubexp(genctx, bn) > 0
-                && EVP_PKEY_CTX_set_rsa_keygen_primes(genctx, primes) > 0
-                && EVP_PKEY_keygen(genctx, &rsa_key);
-            BN_free(bn);
-            bn = NULL;
-            EVP_PKEY_CTX_free(genctx);
-            genctx = NULL;
+            for (i = 0; i < loopargs_len ; ++i)
+            {
+                bn = BN_new();
+                st = bn != NULL
+                    && BN_set_word(bn, RSA_F4)
+                    && init_gen_str(&genctx, "RSA", NULL, 0, NULL, NULL)
+                    && EVP_PKEY_CTX_set_rsa_keygen_bits(genctx, rsa_keys[testnum].bits) > 0
+                    && EVP_PKEY_CTX_set1_rsa_keygen_pubexp(genctx, bn) > 0
+                    && EVP_PKEY_CTX_set_rsa_keygen_primes(genctx, primes) > 0
+                    && EVP_PKEY_keygen(genctx, &rsa_pkeys[i]);
+                BN_free(bn);
+                bn = NULL;
+                EVP_PKEY_CTX_free(genctx);
+                genctx = NULL;
+            }
         } else {
-            const unsigned char *p = rsa_keys[testnum].data;
+           for (i = 0; i < loopargs_len; ++i)
+            {
+                const unsigned char *p = rsa_keys[testnum].data;

-            st = (rsa_key = d2i_PrivateKey(EVP_PKEY_RSA, NULL, &p,
+                st = (rsa_pkeys[i] = d2i_PrivateKey(EVP_PKEY_RSA, NULL, &p,
                                            rsa_keys[testnum].length)) != NULL;
+            }
         }

         for (i = 0; st && i < loopargs_len; i++) {
-            loopargs[i].rsa_sign_ctx[testnum] = EVP_PKEY_CTX_new(rsa_key, NULL);
+            loopargs[i].rsa_sign_ctx[testnum] = EVP_PKEY_CTX_new(rsa_pkeys[i], NULL);
             loopargs[i].sigsize = loopargs[i].buflen;
             if (loopargs[i].rsa_sign_ctx[testnum] == NULL
                 || EVP_PKEY_sign_init(loopargs[i].rsa_sign_ctx[testnum]) <= 0
@@ -2391,7 +2397,7 @@ int speed_main(int argc, char **argv)
         }

         for (i = 0; st && i < loopargs_len; i++) {
-            loopargs[i].rsa_verify_ctx[testnum] = EVP_PKEY_CTX_new(rsa_key,
+            loopargs[i].rsa_verify_ctx[testnum] = EVP_PKEY_CTX_new(rsa_pkeys[i],
                                                                    NULL);
             if (loopargs[i].rsa_verify_ctx[testnum] == NULL
                 || EVP_PKEY_verify_init(loopargs[i].rsa_verify_ctx[testnum]) <= 0
@@ -2424,7 +2430,9 @@ int speed_main(int argc, char **argv)
             /* if longer than 10s, don't do any more */
             stop_it(rsa_doit, testnum);
         }
-        EVP_PKEY_free(rsa_key);
+       for ( i = 0; i < loopargs_len ; ++i )
+            EVP_PKEY_free(rsa_pkeys[i]);
+       free(rsa_pkeys);
     }

     for (testnum = 0; testnum < DSA_NUM; testnum++) {

**Comment by mattcaswell** [comment]
Please could you raise this as a pull request so that it can be more easily reviewed?

**Comment by mattcaswell** [comment]
Anyone can create a pull request:

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request

**Comment by shihyic** [comment]
Hi,
 
Per Mellanox legal, Mellanox is not dissolved into Nvidia and still maintain Mellanox organization. Mellanox legal insists current Mellanox CCLA shall still govern Mellanox developers', including me,  Github/OpenSSL contribution.
 
Please review and merge the patch I submitted.

**Comment by nhorman** [comment]
It appears that the CCLA was not updated to include your name, could you please look into resolving this?

**Comment by nhorman** [comment]
no response, marking as inactive to be closed at the end of 3.4 dev, barring further input

**Comment by shihyic** [comment]
Have been consulting Legal team of Mellanox. They should have talked to OpenSSL team regarding this. Last I know is this from our legal team

“ Mellanox still exists and it’s a legal separate entity than Nvidia Corp. (a subsidiary of Nvidia Corp”
